src/useAsyncSetState.ts

```diff
@@ -0,0 +1,51 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react'
+
+type Updater<TState> = (state: TState) => TState
+
+export type AsyncSetState<TState> = (
+  stateUpdate: React.SetStateAction<TState>,
+) => Promise<TState>
+
+export const useAsyncSetState = <TState>(
+  initialState: TState,
+): [TState, AsyncSetState<TState>] => {
+  const [state, setState] = useState(initialState)
+  const resolvers = useRef<((state: TState) => void)[]>([])
+
+  useEffect(() => {
+    resolvers.current.forEach(resolve => resolve(state))
+    resolvers.current = []
+  }, [state])
+
+  const setStateAsync = useCallback(
+    (update: Updater<TState> | TState) => {
+      return new Promise<TState>((resolve, reject) => {
+        setState(stateBefore => {
```

Comment:
```suggestion
        setState(prevState => {
```

more consistent w/naming otherwise